### PR TITLE
[feature](load) refactor CSV reading process during scanning, and support enclose and escape for stream load

### LINKED ISSUE #22383: [Enhancement] refactor CSV reading process during scanning
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Currently, CSV text format files' reading has the following general keypoints:

1. CsvReader delegates the line reading behavior to NewPlainTextReader.
2. NewPlainTextReader will try to read a whole line from decompressed buffer. If not a whole line is read, it will extend the buffer and read more data, then try to parse a line from the beginning again.
3. When a whole line is read, it will be returned to CsvReader, and CsvReader will try to parse column field from the read line.
4. The process will continue to the end of file, read a line, parse a line, again and again.

However, there are some shortcomings in this way.
1. Low extensibility. When more text attributes are required, such as enclose and escape, it is hard to make change based on the former code.
2. Maybe lower performance. 
    (i) When the line reader doesn't read a whole line, it will try to find the line delimiter again from the beginning, rather than the last position of the former process. 
    (ii) Moreover, when a line is return, the column separators' position is still unknown, and CsvReader will parse the line again from the beginning, and this process could be taken during line reader reading a line.

### Solution

Refactor ideas:

1. Use a state machine to parse a line in l line reader, rather than calling memmem API. This will make parsing process more controllable to developers.
2. Use a context to hold the state of the parser, so that when a whole line is mismatched, the read progress and state will not be discarded so that we can continue the former process.
3. The positions of column separator will be recorded to the context during line reading, and CsvReader can use this info to split the value without parsing the line again.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Hastyshell**
branch is deleted by mistake...... view it at #22539

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

File: be/src/exec/line_reader.h [hunk 1]
```diff
@@ -24,7 +24,7 @@ namespace doris {
 namespace io {
 class IOContext;
 }
-// This class is used for CSV scanner, to read content line by line
+// This class is used to read content line by line
 class LineReader {
```
Comment (yiguolei) [inline be/src/exec/line_reader.h:28]: Split this PR into two PRS:
one pr for refactoring, the other PR for adding support enclose and escape for stream load.

**Comment by TangSiyang2001** [comment]
run buildall

File: be/src/exec/line_reader.h (hunk 1 above)
Comment (TangSiyang2001) [reply be/src/exec/line_reader.h:28]: I'll try to.

**Comment by dataroaring** [comment]
run buildall

File: be/src/exec/line_reader.h [hunk 2]
```diff
@@ -24,7 +24,7 @@ namespace doris {
 namespace io {
 class IOContext;
 }
-// This class is used for CSV scanner, to read content line by line
```
Comment (yiguolei) [inline be/src/exec/line_reader.h:27]: add docs for enclose and escape.

**Comment by TangSiyang2001** [comment]
run buildall

File: be/src/exec/line_reader.h (hunk 2 above)
Comment (TangSiyang2001) [reply be/src/exec/line_reader.h:27]: I'll make it when when it is sure to be correct

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run p0

**Comment by Yukang-Lian** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run click bench

**Comment by TangSiyang2001** [comment]
run clickbench

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
Seems weird, p0 has passed as shown. Correctness is verified, now will resolve the performance problem.
截屏2023-07-31 下午5 40 57

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run clickbench

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run build all

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run clickbench

**Comment by TangSiyang2001** [comment]
run p0

**Pull request closed without merging**